File: kt/main/kotlin/moduleconfig/pushtovar/fwk/CfgSrc.kt
/*
 *  Copyright © 2021 Paulo Villela. All rights reserved.
 *  Use of this source code is governed by the Apache 2.0 license
 *  that can be found in the LICENSE file.
 */

package tryout.moduleconfig.pushtovar.fwk

class CfgSrc<T>(
    private var src: () -> T = ::nilCfgSrc
) {
    fun set(src: () -> T) {
        this.src = src
    }

    operator fun invoke(): T {
        return src()
    }
}

private fun <T>nilCfgSrc(): T {
    throw ConfigurationException("Module used before being initialized")
}

typealias CfgSrcAdapter<S, T> = (S) -> T

class CfgSrcAdaptation<S, T>(
    private val targetSrc: CfgSrc<T>,
    private val adapter: CfgSrcAdapter<S, T>
) {
    fun setOrigin(originSrc: () -> S) {
        this.targetSrc.set { this.adapter(originSrc()) }
    }
}

File: go/module-config/pull-with-push-override/config/cfg-src.go
/*
 *  Copyright © 2021 Paulo Villela. All rights reserved.
 *  Use of this source code is governed by the Apache 2.0 license
 *  that can be found in the LICENSE file.
 */

package config

type CfgSrc[T any] func() T

func nilCfgSrc[T any]() T {
	panic("Module used before being initialized")
}

func MakeCfgSrc[T any](adapter func(src AppCfgInfo) T) CfgSrc[T] {
	if adapter == nil {
		return nilCfgSrc[T]
	}
	return func() T { return adapter(GetAppConfiguration()) }
}
